### ISSUE TITLE
Bug in aggregate subscriptions based on rum derived atoms ?

### ISSUE THREAD

**Issue by DjebbZ** [roman01la/scrum#13]
Hi,

I'm following the README to the letter and I think I found a bug. See the following minimal reproduction case in the REPL :

``` clojure
(require '[scrum.core :as scrum])
=> nil
(require '[rum.core :as rum])
=> nil
(let [rec (scrum/reconciler {:state (atom {})
                             :resolvers {[:a] (constantly :a)}})
      sub (fn [reconciler]
            (scrum/subscription reconciler [:a]))
      der (fn [reconciler]
            (rum/derived-atom [(sub reconciler)] ::key (constantly :derived)))]
  @(der rec))
=> ClassCastException scrum.resolver.Resolver cannot be cast to clojure.lang.IRef  clojure.core/add-watch (core.clj:2134)
```

Whereas plain derived-atoms in rum work fine :

``` clojure
(let [a (atom :a)
      b (atom :b)
      der (rum/derived-atom [a b] ::key
                            (fn [a b] 
                              (println "a" a "b" b)))]
  (reset! a :aaa)
  (reset! b :bbb))
a :a b :b
a :aaa b :b
a :aaa b :bbb
=> :bbb
```

The problem : [rum calls `add-watch` on each ref](https://github.com/tonsky/rum/blob/0.10.8/src/rum/derived_atom.cljc#L26), and in clojure.core [`add-watch` has a type hint on the ref of `clojure.lang.IRef`](https://github.com/clojure/clojure/blob/clojure-1.8.0/src/clj/clojure/core.clj#L2068).

(I think) I was able to track down the root cause : scrum's Resolver only implements `clojure.lang.IDeref` protocol, but [`clojure.lang.IRef`](https://github.com/clojure/clojure/blob/clojure-1.8.0/src/jvm/clojure/lang/IRef.java#L15) extends IDeref and implements more methods. I think the Resolver should implement IRef as well to work just fine. Not sure, I'm not very familiar with Clojure's protocols.

In the meantime, not sure what I should do to work around the problem.

**Comment by DjebbZ** [comment]
As a temporary workaround I'll do the following. What do you think ?

``` clojure
(let [rec   (scrum/reconciler {:state     (atom {})
                               :resolvers {[:a] (constantly :a)
                                           [:b] (constantly :b)}})
      sub-a (fn [reconciler]
              (scrum/subscription reconciler [:a]))
      sub-b (fn [reconciler]
              (scrum/subscription reconciler [:b]))
      derived (fn [reconciler]
                (let [a @(sub-a reconciler)
                      b @(sub-b reconciler)]
                  [a b]))]
  (derived rec))
```

The example is dumb, but I hope you get the point.

**Comment by DjebbZ** [comment]
FYI, not working, since I'm using it as a subscription (with `(rum/react XXX)`)... I think I'm blocked...

**Comment by roman01la** [comment]
Hey. You are right. I'm going to update the code later today. Which version of Scrum is it?

**Comment by DjebbZ** [comment]
`2.1.0-SNAPSHOT`. Sorry I should have mentionned it.

**Comment by roman01la** [comment]
No worries, I'll update both.

**Comment by roman01la** [comment]
Actually I think that's inconsistency in Rum. When used in Clojure derived-atom shouldn't add watchers.

**Comment by DjebbZ** [comment]
Indeed, I think you're right : you render in one pass server-side, watches are useless. At least so it seems. Are you opening an issue in Rum ?

**Comment by roman01la** [comment]
Yeah. Going to talk to Nikita about this.

**Comment by roman01la** [comment]
Fixed in 86d9adf for 2.1.0 and 2.2.9 in master. Please check if it works for you.

**Comment by DjebbZ** [comment]
Looks like it works fine in 2.1.0. Thanks a lot !
